test/health-check.test.js

```diff
@@ -34,28 +36,91 @@ t.test('checkReadiness', async t => {
   })
 
   t.test('should return 503 on S3 error', async t => {
+    reset()
     const awsClient = {
       dynamoQueryBySortKey: async () => { },
       s3Fetch: async () => { throw new Error('ERROR_ON_S3') }
     }
     const logger = helper.spyLogger()
 
+    t.equal(await checkReadiness({ awsClient, readiness, logger }), 200)
     t.equal(await checkReadiness({ awsClient, readiness, logger }), 503)
     t.equal(logger.messages.error.length, 1)
     t.match(logger.messages.error[0][0].err, 'ERROR_ON_S3')
     t.equal(logger.messages.error[0][1], 'Readiness Probe Failed')
   })
 
-  t.test('should return 503 on S3 and Dynamo error', async t => {
+  t.test('should do sampling, without failures', async t => {
+    reset()
+    let dynamoCalls = 0
+    let s3Calls = 0
     const awsClient = {
-      dynamoQueryBySortKey: async () => { throw new Error('ERROR_ON_DYNAMO') },
-      s3Fetch: async () => { throw new Error('ERROR_ON_S3') }
+      dynamoQueryBySortKey: async () => { dynamoCalls++ },
+      s3Fetch: async () => { s3Calls++ }
     }
-    const logger = helper.spyLogger()
+    const logger = helper.dummyLogger()
+    const samplings = [10, 10]
 
-    t.equal(await checkReadiness({ awsClient, readiness, logger }), 503)
-    t.equal(logger.messages.error.length, 1)
-    t.match(logger.messages.error[0][0].err, 'ERROR_ON_DYNAMO')
-    t.equal(logger.messages.error[0][1], 'Readiness Probe Failed')
+    for (let i = 0; i < 100; i++) {
+      t.equal(await checkReadiness({ awsClient, readiness, samplings, logger }), 200)
+    }
+    t.equal(dynamoCalls, 5)
+    t.equal(s3Calls, 5)
+  })
+
+  t.test('should do sampling, resetting on failures', async t => {
+    reset()
+    let dynamoCalls = 0
+    let s3Calls = 0
+    const awsClient = {
+      dynamoQueryBySortKey: async () => {
+        dynamoCalls++
+        if (dynamoCalls === 2) {
+          throw new Error('ERROR_ON_DYNAMO')
+        }
+      },
+      s3Fetch: async () => {
+        s3Calls++
+        if (s3Calls === 3) {
+          throw new Error('ERROR_ON_S3')
+        }
+      }
+    }
+    const logger = helper.dummyLogger()
+    const samplings = [3, 5]
+
+    t.equal(await checkReadiness({ awsClient, readiness, samplings, logger }), 200, '#1')
+    t.equal(await checkReadiness({ awsClient, readiness, samplings, logger }), 200, '#2')
+    t.equal(await checkReadiness({ awsClient, readiness, samplings, logger }), 200, '#3')
+    t.equal(await checkReadiness({ awsClient, readiness, samplings, logger }), 200, '#4')
+    t.equal(await checkReadiness({ awsClient, readiness, samplings, logger }), 200, '#5')
```

Comment:
This test is encoding the behavior I noticed in the implementation - I'd expect this test to expect a failure from S3 only on the 9th readiness check (given we saying make a real check to S3 every 3 calls, and we're saying we should fail on the third real call to S3 via the awsClient).